ampd/src/lib.rs

```diff
@@ -259,23 +271,55 @@ where
             exit_token.cancel();
         });
 
+        let (state_tx, mut state_rx) = oneshot::channel::<State>();
         let mut set = JoinSet::new();
         set.spawn(event_sub.run().change_context(Error::EventSub));
         set.spawn(event_processor.run().change_context(Error::EventProcessor));
         set.spawn(broadcaster.run().change_context(Error::Broadcaster));
-        set.spawn(state_updater.run().change_context(Error::StateUpdater));
+        set.spawn(async move {
+            // assert: the app must wait for this task to exit before trying to receive the state
+            state_tx
+                .send(state_updater.run().await)
+                .expect("the state receiver should still be alive");
+            Ok(())
+        });
 
-        let res = match (set.join_next().await, token.is_cancelled()) {
+        let execution_result = match (set.join_next().await, token.is_cancelled()) {
             (Some(result), false) => {
                 token.cancel();
-                result.change_context(Error::Task)?
+                match result {
+                    Ok(result) => result,
```

Comment:
I know this PR didn't change this part, but now I'm thinking about how could any of the tasks exit with OK if the token isn't cancelled.

Reply:
It's the correct behaviour. It just happens that currently we don't have any tasks that could shut down ampd, so we won't end up in this branch. But I'd rather keep it then break the app in the future because this match is not doing what it's supposed to